cmd/system-probe/subcommands/run/command.go

```diff
@@ -141,25 +141,73 @@ func run(log log.Component, config config.Component, telemetry telemetry.Compone
+//
+// Returns an error channel that can be used to wait for the agent to stop and get the result.
+func StartSystemProbeWithDefaults(ctxChan <-chan context.Context) (<-chan error, error) {
+	errChan := make(chan error)
+
 	// run startSystemProbe in an app, so that the log and config components get initialized
-	return fxutil.OneShot(
-		func(log log.Component, config config.Component, telemetry telemetry.Component, sysprobeconfig sysprobeconfig.Component, rcclient rcclient.Component) error {
-			return startSystemProbe(&cliParams{GlobalParams: &command.GlobalParams{}}, log, telemetry, sysprobeconfig, rcclient)
-		},
-		// no config file path specification in this situation
-		fx.Supply(config.NewAgentParamsWithoutSecrets("", config.WithConfigMissingOK(true))),
-		fx.Supply(sysprobeconfig.NewParams(sysprobeconfig.WithSysProbeConfFilePath(""))),
-		fx.Supply(log.LogForDaemon("SYS-PROBE", "log_file", common.DefaultLogFile)),
-		rcclient.Module,
-		config.Module,
-		telemetry.Module,
-		sysprobeconfig.Module,
-		// use system-probe config instead of agent config for logging
-		fx.Provide(func(lc fx.Lifecycle, params log.Params, sysprobeconfig sysprobeconfig.Component) (log.Component, error) {
-			return log.NewLogger(lc, params, sysprobeconfig)
-		}),
-	)
+	go func() {
+		err := fxutil.OneShot(
+			func(log log.Component, config config.Component, telemetry telemetry.Component, sysprobeconfig sysprobeconfig.Component, rcclient rcclient.Component) error {
+				defer StopSystemProbeWithDefaults()
+				err := startSystemProbe(&cliParams{GlobalParams: &command.GlobalParams{}}, log, telemetry, sysprobeconfig, rcclient)
+				// notify outer that startAgent finished
+				errChan <- err
+				if err != nil {
+					return err
+				}
+				// wait for context
+				ctx := <-ctxChan
+
+				// Wait for stop signal
+				select {
+				case <-signals.Stopper:
+					log.Info("Received stop command, shutting down...")
+				case <-signals.ErrorStopper:
+					_ = log.Critical("The Agent has encountered an error, shutting down...")
+				case <-ctx.Done():
+					log.Info("Received stop from service manager, shutting down...")
+				}
+
+				return nil
+			},
+			// no config file path specification in this situation
+			fx.Supply(config.NewAgentParamsWithoutSecrets("", config.WithConfigMissingOK(true))),
+			fx.Supply(sysprobeconfig.NewParams(sysprobeconfig.WithSysProbeConfFilePath(""))),
+			fx.Supply(log.LogForDaemon("SYS-PROBE", "log_file", common.DefaultLogFile)),
+			rcclient.Module,
+			config.Module,
+			telemetry.Module,
+			sysprobeconfig.Module,
+			// use system-probe config instead of agent config for logging
+			fx.Provide(func(lc fx.Lifecycle, params log.Params, sysprobeconfig sysprobeconfig.Component) (log.Component, error) {
+				return log.NewLogger(lc, params, sysprobeconfig)
+			}),
+		)
+		// notify caller that fx.OneShot is done
+		errChan <- err
+	}()
+
+	// Wait for startSystemProbe to complete, or for an error
+	err := <-errChan
```

Comment:
Can this channel be separate from the one returned to the caller? It was initially confusing to me on the order of operations. Since this one is only used within this function to know when `startSystemProbe` is done, it can be separate.

Reply:
`fxutil.OneShot` can fail before calling the `delayedFunc` that calls `startSystemProbe`. I could separate them, but the commented on code would have to wait on both channels anyway. Would that be more clear?

Reply:
So if `fxutil.OneShot` fails before `delayedFunc`, one value is sent to `errChan`. Otherwise `errChan` gets two values sent, correct (explicit send on 161 and return value from `fxutil.OneShot`).

In the latter case, the goroutine is blocked if an error OR if the caller doesn't consume the second value from the channel?

Reply:
> So if fxutil.OneShot fails before delayedFunc, one value is sent to errChan. Otherwise errChan gets two values sent, correct? (explicit send on 161 and return value from fxutil.OneShot).

correct

> In the latter case, the goroutine is blocked if an error OR if the caller doesn't consume the second value from the channel?

I think the goroutine is blocked only in the case where`startSystemProbe` fails. The process will exit so it's not a big deal. But I think that can be fixed by moving the explicit send on 161 after the `if err != nil` check.

Reply:
Let's do that. Excess goroutines can cause problems in tests.